### PR TITLE
Upgrade to Django 4.2.x and Python 3.12.x

### PR DESCRIPTION
**Pull request by stevejalim**
This changeset moves Nucleus up to Python and Django versions that are not too close to EOL.

The move to Python 3.12.1 also meant moving away from gunicorn+meinheld, as meinheld [won't compile above Python 3.9](https://github.com/mozilla/bedrock/issues/13247)

While in the neighbourhood, I've also switched to `ruff` from `isort` + `flake8`

## To test:

* check out branch
* `make clean build test`
* `docker-compose up web`
* `make run-shell` then `./manage.py createsuperuser` to be able to log in
* Go to http://localhost:8000 and log in to the nucleus admin
* add a release or two, with some notes to confirm all works ok 

## Other considerations:

* We can also test this on nucleus dev once it's hit `master`, but there's no test env. Suggestions/recollections for how this has been tested in the past are welcome.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # BUILDER IMAGE
-FROM python:3.9-slim-bullseye AS builder
+FROM python:3.12-slim-bullseye AS builder
```
Comment (robhudson) [inline Dockerfile:2]: Is there a reason to stay on bullseye and not also bump to the default 3.12 Debian here (bookworm)?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -5,16 +5,12 @@ exclude: >
     | jenkins
   )
 repos:
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: v0.1.11
+    hooks:
+      - id: ruff
   - repo: https://github.com/psf/black
-    rev: 23.10.1
+    rev: 23.12.1
```
Comment (robhudson) [inline .pre-commit-config.yaml:13]: On basket I also dropped black in favor of `ruff format` and it's working well.

File: nucleus/base/admin.py [hunk 3]
```diff
@@ -35,5 +35,4 @@ def has_delete_permission(self, request, obj=None):
         return False
 
 
-admin.site.register(GithubLog, GithubLogAdmin)
 admin.site.register(admin.models.LogEntry, LogEntryAdmin)
```
Comment (robhudson) [inline nucleus/base/admin.py:38]: `django-upgrade` missed this one for me to. You can do it manually so these decorators are consistent.

File: Dockerfile (hunk 1 above)
Comment (stevejalim) [reply Dockerfile:2]: Good call - yes, will update

**Review by robhudson — APPROVED** [review]
Nice 🥳

**Pull request merged**